File: src/main/java/com/joraph/schema/EntityDescriptor.java
package com.joraph.schema;

import static java.util.Objects.requireNonNull;

import java.beans.IntrospectionException;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Predicate;

import com.joraph.CollectionUtil;

import kotlin.jvm.functions.Function1;

/**
 * Metadata about an entity class.
 */
public class EntityDescriptor<T> {

	private final Class<T> entityClass;
	private final Map<Function1<T, ?>, ForeignKey<T, ?>> foreignKeys = new HashMap<>();
	private Class<?> graphKey;
	private Property<T, ?> primaryKey;

	/**
	 * Creates a new instance of EntityDescriptor.
	 * @param entityClass the entity class
	 */
	public EntityDescriptor(Class<T> entityClass) {
		this.entityClass = entityClass;
		this.graphKey = entityClass;
	}

	/**
	 * @return the entityClass
	 */
	public Class<T> getEntityClass() {
		return entityClass;
	}

	/**
	 * @return the primaryKey
	 */
	public Property<T, ?> getPrimaryKey() {
		return primaryKey;
	}

	/**
	 * @return the graphKey
	 */
	public Class<?> getGraphKey() {
		return graphKey;
	}

	/**
	 */
	public <RR> EntityDescriptor<T> setPrimaryKey(Function1<T, RR> fun) {
		this.primaryKey = new Key<>(fun);
		return this;
	}

	/**
	 */
	public EntityDescriptor<T> setGraphKey(Class<?> graphKey) {
		this.graphKey = requireNonNull(graphKey, "graphKey cannot be null");
		return this;
	}

	/**
	 */
	@SafeVarargs
	@SuppressWarnings("unchecked") 
	public final EntityDescriptor<T> setPrimaryKey(
			Function1<Object[], ?> converter,
			Function1<T, ?> first,
			Function1<T, ?>... remaining) {
		Function1<T, ?>[] chains = CollectionUtil.asStream(first, remaining)
				.toArray(Function1[]::new);
		this.primaryKey = new CompositeKey<>(converter, chains);
		return this;
	}

	/**
	 */
	@SafeVarargs
	public final EntityDescriptor<T> setPrimaryKey(Function1<T, ?> first, Function1<T, ?>... remaining) {
		return setPrimaryKey(BasicCompositeKey.CONVERTER, first, remaining);
	}
 
	/**
	 * @return the foreignKeys
	 */
	public Map<Function1<?, ?>, ForeignKey<?, ?>> getForeignKeys() {
		return Collections.unmodifiableMap(foreignKeys);
	}

	/**
	 * Creates a builder for adding a foreign key.
	 */
	public ForeignKeyBuilder<?, T> addForeignKey(Class<?> foreignEntity) {
		return new ForeignKeyBuilder<>(this, foreignEntity);
	}

	/**
	 * Adds a foreign key.
	 */
	public EntityDescriptor<T> addForeignKey(Class<?> foreignEntity, Function1<T, ?> accessor) {
		addForeignKey(foreignEntity, accessor, null, null);
		return this;
	}

	/**
	 * Adds a foreign key.
	 */
	public <A> EntityDescriptor<T> addForeignKey(Class<?> foreignEntity, Function1<T, ?> accessor, Class<A> argumentClass, Predicate<A> argumentPredicate)  {
		this.foreignKeys.put(accessor, new ForeignKey<>(entityClass, foreignEntity, argumentClass, argumentPredicate, accessor));
		return this;
	}

	/**
	 * Returns a {@link ForeignKey} by name.
	 */
	public ForeignKey<T, ?> getForeignKey(Function1<Object, ?> accessor) {
		return foreignKeys.get(accessor);
	}


	public class ForeignKeyBuilder<A, TT> {

		private final EntityDescriptor<TT> entity;
		private final Class<?> foreignEntity;
		private Function1<TT, ?> accessor;
		private Class<A> argumentClass;
		private Predicate<A> argumentPredicate;

		private ForeignKeyBuilder(EntityDescriptor<TT> entity, Class<?> foreignEntity) {
			this.entity = entity;
			this.foreignEntity = foreignEntity;
		}

		public ForeignKeyBuilder<A, TT> withAccessor(Function1<TT, ?> accessor) {
			this.accessor = accessor;
			return this;
		}

		@SuppressWarnings("unchecked")
		public <A2> ForeignKeyBuilder<A2, TT> withPredicate(Class<A2> argumentClass, Predicate<A2> argumentPredicate) {
			this.argumentClass = (Class<A>)argumentClass;
			this.argumentPredicate = (Predicate<A>)argumentPredicate;
			return (ForeignKeyBuilder<A2, TT>)this;
		}

		public EntityDescriptor<TT> add()
				throws IntrospectionException {
			return entity.addForeignKey(foreignEntity, accessor, argumentClass, argumentPredicate);
		}

	}

}

File: src/main/kotlin/com/joraph/schema/EntityDescriptor.kt
package com.joraph.schema

import java.util.*
import java.util.function.Predicate


/**
 * Metadata about an entity class.
 */
class EntityDescriptor<T> @JvmOverloads constructor(
        val entityClass: Class<T>,
        primaryKey: Key<T, *>? = null,
        var graphKey: Class<*> = entityClass) {

    val foreignKeys: MutableMap<Function1<*, *>, ForeignKey<*, *>> = mutableMapOf()

    @JvmOverloads constructor(
            entityClass: Class<T>,
            keyFunction: Function1<T, *>,
            graphKey: Class<*> = entityClass) : this(
            entityClass = entityClass,
            primaryKey = Key(keyFunction),
            graphKey = graphKey)

    private var _primaryKey: Key<T, *>? = primaryKey

    val primaryKey: Key<T, *> get() = _primaryKey
            ?: throw MissingPrimaryKeyException(this)

    fun withGraphKey(graphKey: Class<*>): EntityDescriptor<T> {
        this.graphKey = graphKey
        return this
    }

    fun withPrimaryKey(func: Function1<T, *>): EntityDescriptor<T> {
        _primaryKey = Key(func)
        return this
    }

    @SafeVarargs
    fun withPrimaryKey(
            cconverter: Function1<Array<Any?>, *>,
            first: Function1<T, *>,
            vararg accessors: Function1<T, *>): EntityDescriptor<T> {
        _primaryKey = SchemaUtil.compositeKey(cconverter, first, *accessors)
        return this
    }

    @SafeVarargs
    fun withPrimaryKey(first: Function1<T, *>, vararg remaining: Function1<T, *>): EntityDescriptor<T> {
        return withPrimaryKey(BasicCompositeKey.CONVERTER, first, *remaining)
    }

    fun addForeignKey(foreignEntity: Class<*>): ForeignKeyBuilder<*, T> {
        return ForeignKeyBuilder<Any, T>(this, foreignEntity)
    }

    fun withForeignKey(foreignEntity: Class<*>, accessor: Function1<T, *>): EntityDescriptor<T> {
        foreignKeys[accessor] = ForeignKey(entityClass, foreignEntity, accessor)
        return this
    }

    fun <A> withForeignKey(foreignEntity: Class<*>, accessor: Function1<T, *>, argumentClass: Class<A>, argumentPredicate: Predicate<A>): EntityDescriptor<T> {
        foreignKeys[accessor] = ConditionalForeignKey(entityClass, foreignEntity, accessor, argumentClass, argumentPredicate)
        return this
    }

    inner class ForeignKeyBuilder<A, TT> internal constructor(
            private val entity: EntityDescriptor<TT>,
            private val foreignEntity: Class<*>) {

        private var accessor: Function1<TT, *>? = null
        private var argumentClass: Class<A>? = null
        private var argumentPredicate: Predicate<A>? = null

        fun withAccessor(accessor: Function1<TT, *>?): ForeignKeyBuilder<A, TT> {
            this.accessor = accessor
            return this
        }

        @Suppress("UNCHECKED_CAST")
        fun <A2> withPredicate(argumentClass: Class<A2>, argumentPredicate: Predicate<A2>): ForeignKeyBuilder<A2, TT> {
            this.argumentClass = argumentClass as Class<A>?
            this.argumentPredicate = argumentPredicate as Predicate<A>?
            return this as ForeignKeyBuilder<A2, TT>
        }

        fun add(): EntityDescriptor<TT> {
            val acc = accessor
                    ?: throw IllegalStateException("accessor is null")

            return if (argumentClass == null || argumentPredicate == null) {
                entity.withForeignKey(foreignEntity, acc)
            } else {
                entity.withForeignKey(foreignEntity, acc, argumentClass!!, argumentPredicate!!)
            }
        }
    }

    /**
     * Creates a new instance of EntityDescriptor.
     * @param entityClass the entity class
     */
    init {
        graphKey = entityClass
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as EntityDescriptor<*>

        if (entityClass != other.entityClass) return false
        if (graphKey != other.graphKey) return false
        if (foreignKeys != other.foreignKeys) return false
        if (_primaryKey != other._primaryKey) return false

        return true
    }

    override fun hashCode(): Int {
        var result = entityClass.hashCode()
        result = 31 * result + graphKey.hashCode()
        result = 31 * result + foreignKeys.hashCode()
        result = 31 * result + (_primaryKey?.hashCode() ?: 0)
        return result
    }

}
